src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs

```diff
@@ -0,0 +1,427 @@
+                Model = null;
+                FilePath = null;
+                _lastWriteTime = DateTime.MinValue;
+            }
+        }
+
+        private class ProjectContextEntry
+        {
+            public List<ProjectContext> ProjectContexts { get; } = new List<ProjectContext>();
+
+            public string LockFilePath { get; set; }
+
+            public string ProjectFilePath { get; set; }
+
+            public DateTime LastProjectFileWriteTime { get; set; }
+
+            public DateTime LastLockFileWriteTime { get; set; }
+
+            public bool HasChanged
+            {
+                get
+                {
+                    if (ProjectFilePath == null || !File.Exists(ProjectFilePath))
+                    {
+                        return true;
+                    }
+
+                    if (LastProjectFileWriteTime < File.GetLastWriteTime(ProjectFilePath))
+                    {
+                        return true;
+                    }
+
+                    if (LockFilePath == null || !File.Exists(LockFilePath))
+                    {
+                        return true;
+                    }
+
+                    if (LastLockFileWriteTime < File.GetLastWriteTime(LockFilePath))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                }
+            }
+
+            public void Reset()
+            {
+                ProjectContexts.Clear();
+                ProjectFilePath = null;
+                LockFilePath = null;
+                LastLockFileWriteTime = DateTime.MinValue;
+                LastProjectFileWriteTime = DateTime.MinValue;
+            }
+        }
+
+        private static string NormalizeProjectPath(string path)
+        {
+            if (File.Exists(path) &&
+                string.Equals(Path.GetFileName(path), Project.FileName, StringComparison.OrdinalIgnoreCase))
+            {
+                return Path.GetFullPath(Path.GetDirectoryName(path));
+            }
+            else if (Directory.Exists(path) &&
+                     File.Exists(Path.Combine(path, Project.FileName)))
+            {
+                return Path.GetFullPath(path);
+            }
+
+            return null;
+        }
+
+        private static List<string> ResolveProjectPath(string projectPath)
+        {
+            if (File.Exists(projectPath))
+            {
+                var filename = Path.GetFileName(projectPath);
+                if (!Project.FileName.Equals(filename, StringComparison.OrdinalIgnoreCase) &&
+                    !GlobalSettings.FileName.Equals(filename, StringComparison.OrdinalIgnoreCase))
+                {
+                    return null;
+                }
+
+                projectPath = Path.GetDirectoryName(projectPath);
+            }
+
+            if (File.Exists(Path.Combine(projectPath, Project.FileName)))
+            {
+                return new List<string> { projectPath };
```

Comment:
I thought we were going to have `WorkspaceContext` always look for the `global.json` and lazily locate projects on-demand?